### PR TITLE
suppprt pandas categorical

### PR DESCRIPTION
**Pull request by wxchan**
#192 

also support categorical str

### CODE REVIEW

**Comment by wxchan** [comment]
should I change default to 'auto'?

**Comment by guolinke** [comment]
I think you can. 
BTW, how to ensure the training data and validation data use same code of categorical ?

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -454,9 +454,17 @@ def __pred_for_csc(self, csc, num_iteration, predict_type):
...
     if isinstance(data, DataFrame):
         cat_cols = data.select_dtypes(include=['category']).columns
+        if not pandas_categorical:  # train dataset
+            pandas_categorical = (data[col].cat.categories for col in cat_cols)
+        else:
+            if len(cat_cols) != len(pandas_categorical):
+                raise ValueError('train and valid dataset categorical_feature do not match.')
+            for col, category in zip(cat_cols, pandas_categorical):
+                if data[col].cat.categories != category:
+                    data[col].cat.set_categories(category, inplace=True)
```
Comment (wxchan) [inline python-package/lightgbm/basic.py]: @guolinke I test set_categories will make category codes same. You can give some more tests.

File: python-package/lightgbm/basic.py [hunk 2]
```diff
@@ -779,6 +788,7 @@ def create_valid(self, data, label=None, weight=None, group=None,
                       weight=weight, group=group, silent=silent, params=params,
                       free_raw_data=self.free_raw_data)
         ret._set_predictor(self._predictor)
```
Comment (wxchan) [inline python-package/lightgbm/basic.py]: @guolinke why uses _set_predictor here and ._predictor = ... in subset()? The result should be the same?

File: python-package/lightgbm/basic.py (hunk 1 above)
Comment (guolinke) [inline python-package/lightgbm/basic.py]: @wxchan I think you can add this test

File: python-package/lightgbm/basic.py (hunk 2 above)
Comment (guolinke) [inline python-package/lightgbm/basic.py]: @wxchan you can make them unify. It should be same.

**Comment by wxchan** [comment]
strange thing: add this test will lead other tests fail.

@guolinke I am not sure why test fails, do boosters in same script share the same handle?

**Comment by guolinke** [comment]
@wxchan  can you pass this test offline ?

**Comment by wxchan** [comment]
@guolinke No, same error.

**Comment by guolinke** [comment]
@wxchan 
I try some tests, it seems the ```params``` in ```test_template``` is changed after call ```test_pandas_categorical```, and this params caused error. 
Did you know why ?

**Comment by guolinke** [comment]
@wxchan if I change ```test_regression``` as following:
```
    def test_regreesion(self):
        evals_result, ret = test_template({'objective': 'regression', 'metric': 'l2'})
        ret **= 0.5
        self.assertLess(ret, 4)
        self.assertAlmostEqual(min(evals_result['eval']['l2']), ret, places=5)
```
it can pass...

**Comment by guolinke** [comment]
@wxchan what is the reason?

**Comment by wxchan** [comment]
The params is used as a global variable, I put it into a class.

There is still a warning, I test it occurs with set_categories, but I am not sure how to fix it.

**Comment by wxchan** [comment]
fixed. in order not to change origin data, make a copy of origin data and change category to int, but it will add memory cost.

another issue here: pandas feature name can be integer, so cannot use them directly sometime.

**Pull request merged**